File: packages/shared-types/src/domain/forecast.ts
// Forecast domain types - ML predictions

import type { TenantEntity, UUID, ISODateString } from "../utils/common";
import type { DateRange, TimeGranularity } from "../utils/dates";

/** Forecast model type */
export type ForecastModelType =
  | "arima" // Time series
  | "prophet" // Facebook Prophet
  | "random_forest" // ML ensemble
  | "xgboost" // Gradient boosting
  | "ensemble"; // Combined models

/** Forecast confidence level */
export type ConfidenceLevel = "low" | "medium" | "high";

/** Forecast status */
export type ForecastStatus = "pending" | "running" | "completed" | "failed";

/** Forecast run entity */
export interface ForecastRun extends TenantEntity {
  /** Forecast horizon start */
  horizonStart: ISODateString;
  /** Forecast horizon end */
  horizonEnd: ISODateString;
  /** Granularity */
  granularity: TimeGranularity;
  /** Model used */
  modelType: ForecastModelType;
  /** Model version */
  modelVersion: string;
  /** Status */
  status: ForecastStatus;
  /** Training data range */
  trainingDataRange: DateRange;
  /** Metrics */
  metrics?: ForecastMetrics;
  /** Run started at */
  startedAt: ISODateString;
  /** Run completed at */
  completedAt?: ISODateString;
  /** Error message if failed */
  errorMessage?: string;
  /** Department scope (null = all) */
  departmentId?: UUID;
}

/** Forecast accuracy metrics */
export interface ForecastMetrics {
  /** Mean Absolute Error */
  mae: number;
  /** Mean Absolute Percentage Error */
  mape: number;
  /** Root Mean Square Error */
  rmse: number;
  /** R-squared */
  r2: number;
  /** Coverage of prediction intervals */
  coverageRate: number;
}

/** Daily forecast prediction */
export interface DailyForecast {
  /** Forecast run ID */
  forecastRunId: UUID;
  /** Prediction date */
  date: ISODateString;
  /** Department (null = organization-wide) */
  departmentId?: UUID;
  /** Predicted absence count */
  predictedAbsences: number;
  /** Predicted absence rate (%) */
  predictedAbsenceRate: number;
  /** Lower bound (confidence interval) */
  lowerBound: number;
  /** Upper bound (confidence interval) */
  upperBound: number;
  /** Confidence level */
  confidence: ConfidenceLevel;
  /** Breakdown by absence type */
  byType: Record<string, number>;
  /** Breakdown by category */
  byCategory: Record<string, number>;
  /** Risk indicators */
  riskIndicators: RiskIndicators;
}

/** Risk indicators for a forecast day */
export interface RiskIndicators {
  /** Understaffing risk (0-100) */
  understaffingRisk: number;
  /** Operational impact score (0-100) */
  operationalImpact: number;
  /** Critical roles at risk */
  criticalRolesAtRisk: number;
  /** Departments at risk */
  departmentsAtRisk: UUID[];
  /** Risk level */
  riskLevel: "low" | "medium" | "high" | "critical";
}

/** Forecast summary for a period */
export interface ForecastSummary {
  period: DateRange;
  granularity: TimeGranularity;
  /** Average predicted absence rate */
  avgAbsenceRate: number;
  /** Total predicted absence days */
  totalAbsenceDays: number;
  /** Peak absence date */
  peakDate: ISODateString;
  /** Peak absence count */
  peakCount: number;
  /** High risk days count */
  highRiskDaysCount: number;
  /** Confidence level */
  overallConfidence: ConfidenceLevel;
  /** Daily forecasts */
  dailyForecasts: DailyForecast[];
}

/** Forecast vs Actual comparison */
export interface ForecastAccuracy {
  date: ISODateString;
  predicted: number;
  actual: number;
  difference: number;
  percentageError: number;
  withinConfidenceInterval: boolean;
}

/** Backtesting results */
export interface BacktestResults {
  period: DateRange;
  modelType: ForecastModelType;
  metrics: ForecastMetrics;
  dailyAccuracy: ForecastAccuracy[];
  /** Best performing periods */
  bestPeriods: DateRange[];
  /** Worst performing periods */
  worstPeriods: DateRange[];
  /** Recommendations */
  recommendations: string[];
}

/** Feature importance for ML models */
export interface FeatureImportance {
  feature: string;
  importance: number;
  description: string;
}

/** Model explanation */
export interface ModelExplanation {
  modelType: ForecastModelType;
  featureImportances: FeatureImportance[];
  seasonalPatterns: SeasonalPattern[];
  trendDirection: "increasing" | "decreasing" | "stable";
  confidenceFactors: string[];
}

/** Seasonal pattern detected */
export interface SeasonalPattern {
  type: "weekly" | "monthly" | "yearly" | "holiday";
  description: string;
  impact: number; // multiplier
  peakPeriods: string[];
}

/** Forecast request parameters */
export interface ForecastRequest {
  horizonDays: number;
  granularity: TimeGranularity;
  modelType?: ForecastModelType;
  departmentId?: UUID;
  includeConfidenceIntervals?: boolean;
}

/** What-if scenario for forecasting */
export interface WhatIfScenario {
  name: string;
  description: string;
  /** Absence rate modifier (e.g., 1.2 = 20% increase) */
  absenceRateModifier?: number;
  /** Specific type modifiers */
  typeModifiers?: Record<string, number>;
  /** Additional planned absences */
  additionalAbsences?: {
    employeeId: UUID;
    startDate: ISODateString;
    endDate: ISODateString;
    type: string;
  }[];
}

/** What-if scenario result */
export interface WhatIfResult {
  scenario: WhatIfScenario;
  baselineForecast: ForecastSummary;
  scenarioForecast: ForecastSummary;
  impact: {
    absenceRateChange: number;
    additionalRiskDays: number;
    costImpact?: number;
  };
}

File: app-api/app/schemas/forecast.py
"""Forecast schemas — request/response models.

Maps to shared-types: ForecastRun, DailyForecast, ForecastMetrics.
Also includes compatibility models used by legacy frontend endpoints:
- ForecastSummary
- WhatIfResult
"""

import uuid
from datetime import date, datetime
from typing import Any, Literal

from pydantic import ConfigDict, Field

from app.models.daily_forecast import ForecastDimension
from app.models.forecast_run import ForecastModelType, ForecastStatus
from app.schemas.base import CamelModel, TenantEntitySchema


class ForecastRunRead(TenantEntitySchema):
    """Full forecast run response."""

    model_type: ForecastModelType
    model_version: str | None = None
    horizon_days: int
    status: ForecastStatus
    started_at: datetime | None = None
    completed_at: datetime | None = None
    accuracy_score: float | None = None
    error_message: str | None = None
    department_id: uuid.UUID | None = None
    config: dict[str, Any]


class ForecastRunSummary(CamelModel):
    """Lightweight forecast run for listings."""

    id: uuid.UUID
    model_type: ForecastModelType
    horizon_days: int
    status: ForecastStatus
    accuracy_score: float | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None


class DailyForecastRead(TenantEntitySchema):
    """Full daily forecast response."""

    forecast_run_id: uuid.UUID
    department_id: uuid.UUID | None = None
    forecast_date: date
    dimension: ForecastDimension
    predicted_demand: float
    predicted_capacity: float
    capacity_planned_current: float
    capacity_planned_predicted: float
    capacity_optimal_predicted: float
    gap: float
    risk_score: float
    confidence_lower: float
    confidence_upper: float
    details: dict[str, Any]


class ForecastRequest(CamelModel):
    """Request a new forecast run."""

    model_config = ConfigDict(extra="forbid")

    horizon_days: int
    model_type: ForecastModelType | None = None
    department_id: uuid.UUID | None = None


class ForecastPeriod(CamelModel):
    """Date range for forecast summary periods."""

    start_date: date
    end_date: date


class ForecastRiskIndicators(CamelModel):
    """Risk indicators compatible with shared frontend ForecastSummary types."""

    understaffing_risk: float
    operational_impact: float
    critical_roles_at_risk: int
    departments_at_risk: list[uuid.UUID] = Field(default_factory=list)
    risk_level: Literal["low", "medium", "high", "critical"]


class DailyForecastCompatRead(CamelModel):
    """Frontend-compatible daily forecast shape (legacy contract)."""

    forecast_run_id: uuid.UUID
    date: date
    department_id: uuid.UUID | None = None
    predicted_absences: float
    predicted_absence_rate: float
    lower_bound: float
    upper_bound: float
    confidence: Literal["low", "medium", "high"]
    by_type: dict[str, float] = Field(default_factory=dict)
    by_category: dict[str, float] = Field(default_factory=dict)
    risk_indicators: ForecastRiskIndicators


class ForecastSummaryRead(CamelModel):
    """Frontend-compatible forecast summary shape (legacy contract)."""

    period: ForecastPeriod
    granularity: Literal["day"] = "day"
    avg_absence_rate: float
    total_absence_days: float
    peak_date: date | None = None
    peak_count: float
    high_risk_days_count: int
    overall_confidence: Literal["low", "medium", "high"]
    daily_forecasts: list[DailyForecastCompatRead]


class WhatIfAdditionalAbsence(CamelModel):
    """Additional planned absences passed to what-if scenarios."""

    model_config = ConfigDict(extra="forbid")

    employee_id: uuid.UUID
    start_date: date
    end_date: date
    type: str = Field(min_length=1, max_length=100)


class WhatIfScenarioRequest(CamelModel):
    """What-if input payload compatible with shared frontend types."""

    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    absence_rate_modifier: float | None = Field(default=None, ge=0)
    type_modifiers: dict[str, float] | None = None
    additional_absences: list[WhatIfAdditionalAbsence] | None = None


class WhatIfScenarioRead(CamelModel):
    """Echoed what-if scenario after server-side normalization."""

    name: str
    description: str | None = None
    absence_rate_modifier: float | None = None
    type_modifiers: dict[str, float] | None = None
    additional_absences: list[WhatIfAdditionalAbsence] = Field(default_factory=list)


class WhatIfImpactRead(CamelModel):
    """Delta between baseline and scenario forecasts."""

    absence_rate_change: float
    additional_risk_days: int
    cost_impact: float | None = None


class WhatIfResultRead(CamelModel):
    """Full what-if response payload compatible with frontend shared types."""

    scenario: WhatIfScenarioRead
    baseline_forecast: ForecastSummaryRead
    scenario_forecast: ForecastSummaryRead
    impact: WhatIfImpactRead
